### PR TITLE
32 br host makes couch available

### LINKED ISSUE #32: Host can make a couch available
As a host with a listed couch
when I visit that couch listing page
and I click 'Add Availability'
and I select the dates I want to make available
and I click 'Make Available'
then I should be back on the listing page
and I should see that my couch is available on those dates.

### PR DESCRIPTION
**Pull request by riverswb**
@ski-climb @Sh1pley @njgheorghita @tmikeschu 
Hosts can create nights by adding availability to an existing couch.  For now, the availability just shows on the couch show page so that I could test it.  Not sure how we ultimately want it.  

Closes #32

### CODE REVIEW

File: spec/features/authentication/visitor_creates_an_account_spec.rb [hunk 1]
```diff
@@ -1,43 +1,47 @@
...
+
       expect(current_path).to eq(user_path(user.id))
       expect(page).not_to have_content("Create Account")
       expect(page).to have_content("First Name")
       expect(page).to have_content("Last Name")
-      expect(page).to have_content("Email")
-      expect(page).to have_content("Phone")
+      expect(page).to have_content(email)
+      expect(page).to have_content(phone_number)
     end
```
Comment (riverswb) [inline spec/features/authentication/visitor_creates_an_account_spec.rb:31]: I did this to get a unique email and phone number before the databse cleaner was setup, I can remove it if we want

File: spec/models/couch_spec.rb [hunk 2]
```diff
@@ -32,11 +32,9 @@
     describe ".search()" do
       before do
         couch_1, couch_2, couch_3 = Couch.all[0..2]
-        couch_1.nights << create(:night, date: Date.yesterday)
         couch_1.nights << create(:night, date: Date.current)
         couch_1.nights << create(:night, date: Date.tomorrow)
         couch_2.nights << create(:night, date: Date.current + 10.days)
-        couch_3.nights << create(:night, date: Date.yesterday)
```
Comment (riverswb) [inline spec/models/couch_spec.rb:39]: With new validations nights can't be created with a date in the past

File: spec/rails_helper.rb [hunk 3]
```diff
@@ -71,4 +71,14 @@
   config.filter_rails_from_backtrace!
   # arbitrary gems may also be filtered via:
   # config.filter_gems_from_backtrace("gem name")
+  DatabaseCleaner.strategy = :truncation
```
Comment (riverswb) [inline spec/rails_helper.rb:74]: Should I remove this? Database cleaner wasn't working for me before adding these changes

File: spec/factories/nights.rb [hunk 4]
```diff
@@ -1,7 +1,7 @@
 FactoryGirl.define do
   factory :night do
-    date { Faker::Date.between(2.weeks.ago, 2.weeks.from_now) }
+    date { Faker::Date.between(Date.today, 2.weeks.from_now) }
```
Comment (riverswb) [inline spec/factories/nights.rb:3]: Dates can't be in past now

**Comment by ski-climb** [comment]
I think we want to add our `/public` directory to the `.gitignore`, otherwise, it looks like we'll all be committing a bunch of uploaded couch images, and worse, making them part of this git project.

**Comment by riverswb** [comment]
@ski-climb  Good call, done

File: spec/features/authentication/visitor_creates_an_account_spec.rb (hunk 1 above)
Comment (ski-climb) [reply spec/features/authentication/visitor_creates_an_account_spec.rb:31]: Would we be able to use the factories too?

Comment (riverswb) [reply spec/features/authentication/visitor_creates_an_account_spec.rb:31]: Is there a way with Factory girl to make a user with attributes but not save?  We could use Faker::Internet.email & and Faker::PhoneNumber.phone_number

File: .gitignore [hunk 5]
```diff
@@ -21,3 +21,6 @@
 
 # Ignore simplecov
 /coverage
+
+# Ignore public Directory
+/public
```
Comment (tmikeschu) [inline .gitignore:26]: Is this because of image uploads?

File: app/controllers/couches/nights_controller.rb [hunk 6]
```diff
@@ -0,0 +1,34 @@
+class Couches::NightsController < ApplicationController
+
+  def new
+    @night = Night.new
+  end
+
+  def create
+    couch = Couch.find(params[:couch_id])
+    begin
+      Night.transaction do
```
Comment (tmikeschu) [inline app/controllers/couches/nights_controller.rb:10]: Cool! Would love a walkthrough on this.

File: app/controllers/couches/nights_controller.rb [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+
+      rescue ActiveRecord::RecordInvalid => invalid
+        flash[:danger] = invalid.message.split(':')[1].strip
+        redirect_to new_couch_night_path(couch)
+    end
+  end
+
+  private
+    def night_params
+      params.permit("First Night", "Last Night")
```
Comment (tmikeschu) [inline app/controllers/couches/nights_controller.rb:26]: We can refactor this once we get @ski-climb 's formatting logic in.

File: app/models/night.rb [hunk 8]
```diff
@@ -1,10 +1,17 @@
...
   belongs_to :couch
-  belongs_to :reservation
+  belongs_to :reservation, optional: true
 
   scope :all_for_couch, -> (couch) { where(couch: couch) }
   scope :between_check_in_check_out, -> (check_in, check_out) {
     where(date: (Date.parse(check_in))..(Date.parse(check_out) - 1.day))
   }
+
+  def night_cannot_be_in_the_past
```
Comment (tmikeschu) [inline app/models/night.rb:12]: Nice custom validation!

File: spec/features/authentication/visitor_creates_an_account_spec.rb (hunk 1 above)
Comment (tmikeschu) [reply spec/features/authentication/visitor_creates_an_account_spec.rb:31]: @riverswb what do you mean not save?

File: spec/rails_helper.rb (hunk 3 above)
Comment (tmikeschu) [reply spec/rails_helper.rb:74]: @riverswb if you needed it for the tests to work, I'd say leave it.

**Review by tmikeschu — COMMENTED** [review]
Is it possible to remove those public image uploads?

Otherwise, looks solid. Nice work!

File: spec/features/authentication/visitor_creates_an_account_spec.rb (hunk 1 above)
Comment (riverswb) [reply spec/features/authentication/visitor_creates_an_account_spec.rb:31]: This is for visitor creating an account, so if the factory creates the account the validations will fail when a user tries to create an account with the same info, right?

File: .gitignore (hunk 5 above)
Comment (riverswb) [reply .gitignore:26]: Yes, this should keep images from being included in git

File: spec/features/authentication/visitor_creates_an_account_spec.rb (hunk 1 above)
Comment (tmikeschu) [reply spec/features/authentication/visitor_creates_an_account_spec.rb:31]: Ohhhh I see what you're saying. I think making an existing user with Faker/Factory then just doing the fill_in steps is a fine way to go.

File: .gitignore (hunk 5 above)
Comment (tmikeschu) [reply .gitignore:26]: @riverswb sorry I didn't read above between you and Nick

File: spec/features/authentication/visitor_creates_an_account_spec.rb (hunk 1 above)
Comment (ski-climb) [reply spec/features/authentication/visitor_creates_an_account_spec.rb:31]: Or `user = build(:user)` will make an instance but not save it to the database.

**Pull request merged**